### ISSUE TITLE
Add map scale feature

### ISSUE THREAD

**Issue by schwukas** [hholzgra/maposmatic#23]
Well, this is kind of self explanatory. I think being able to choose a specfic map scale would greatly increase usability.

As I have zero to no experience with cartography I don't think I can help a whole lot on that addition, but nonetheless I try to help wherever I can.

**Comment by Skylab56** [comment]
Not sure if this is any help, but noticed that www.milvusmaps.eu has ability to choose a specific map scale. Also agree this would be a welcome addition. Sorry I'm a novice, but happy to help if I can.

**Comment by maltfield** [comment]
The ability to have a user define the scale of the maps generated has been subsequently asked on these two issues:

 * https://github.com/hholzgra/maposmatic/issues/61
 * https://github.com/hholzgra/maposmatic/issues/74

+1 for this feature, but here's a workaround:

1. Select some area on the map and click `Next`
1. Select `Multi-page layout` and click `Next`
2. Choose your stylesheet (I prefer `OpenTopMap`) and click `Next`
3. Enable the `Scale bar` Overlay
4. (optional) enable the `Contour lines at 100m resolution Overaly
5. Click `Next`
6. Choose your paper size and click `Next`
7. Give your map a name and click `Generate`

The resulting map will be generated and tell you the scale. Personally, I'm shooting for a scale between `1:50,000` to `1:100,000`

If it's too zoomed-in (eg scale says `1:<something less than 50,000>`), then go back to step 1 above, but select a *bigger* area.

If it's too zoomed-out ((eg scale says `1:<something less than 50,000>`) then go back to step 1 above, but select a "smaller" area.

Usually I just have this issue when I select too small of an area. So I select a bigger area until the desired scale is achieved, and print a subset of the pages in the PDF.

**Comment by Talkless** [comment]
+1 for this feature. Some compasses (like Silva Expedition, etc) has specialized "rulers" on them with standard scales like 1:25000 or 1:50000 for measuring distances more conveniently.

**Comment by danielhjames** [comment]
> Not sure if this is any help, but noticed that [www.milvusmaps.eu](http://www.milvusmaps.eu) has ability to choose a specific map scale.

Sorry to say this site is now offline.

**Comment by 0x416e6f6e** [comment]
This is a must-have feature. This is absolutely necessary for printable maps. It's astounding that the project has existed for so long without such an essential, fundamental feature.
I think this should be a high priority. Is anyone working on this now?

**Comment by ColinChampion** [comment]
I too am surprised that there isn’t a standard tool for producing large format maps from OSM data, but it isn’t MyOSMatic’s fault that it doesn’t seek to be one. 
For my own purposes I made a modified interface [printmap](https://www.routemaster.app/printmap) for invoking the MyOSMatic back end. It’s a bit clunky. It allows you to input bounding lat and long and a scale instead of a paper size. Even if I’ve correctly understood the relationship between the parameters the resulting scale will not be exact because the paper size is required to be a whole number of mm. 
It follows that if you create two maps at different latitudes at a nominal scale of 1:100000, then one will come out at 1:100100 and the other at 1:99900; and since 1:100000 is the boundary between two zooms, they might have different levels of detail. A workaround would be to produce maps at 1:90909 or 1:111111 and change the dpi from 300 to 330 or 270. Perhaps in an ideal world the zoom would be separately specifiable. 
I haven’t found a style sheet I was happy with. Often roads and streams are too pale and thin to be visible; in the orientation styles, on the contrary, place names are suppressed. Only Veloroad is satisfactory for visibility, but like most styles it suppresses too many details (including sub-100m contours) at 1:100k. OpenTopoMap preserves contours better, but its hill shading seriously degrades legibility and it doesn’t show enough place names. Perhaps I could do better if I could request a zoom of 14 at a scale of 1:100k.
A problem with the UTM grid is that the coordinates are shown _inside_ the map; it is better for them to lie outside the edge (on all 4 sides). This isn’t a natural option to provide in MyOSMatic, which treats the grid as an overlay and constrains overlays to lie within the map canvas. It would probably be possible to write a postprocessor which reframed the map with a correctly labelled grid. 
It would be nice to have 20m or 25m as a contour option.

**Comment by danielhjames** [comment]
Nice work @ColinChampion ! I tested generating a map at 1:1250 and got 1:1260, which although not quite there is encouraging. Perhaps the paper size can be padded to get to a whole number of millimetres?

The way we usually deal with this using other tools is just to have a white border around the frame of the map itself; because the extent of the map is so flexible, it never fits into a standard paper size exactly at a fixed scale.

**Comment by ColinChampion** [comment]
@danielhjames: that’s an error of about 1% in the scale, which is much the same as I was seeing myself. The paper padding would account for about a tenth of that, so there must be something else going wrong. I’m afraid I cannot sufficiently understand the Ocitysmap code to work out what it might be :-(

**Comment by ColinChampion** [comment]
Well, if anyone’s interested, I wrote a postprocessor to present OCitysMap output in a layout more suitable for large format printing. My test map, an area in southern Peru, is [here](https://www.routemaster.app/software/images/ausangate.png) (56MB). The printing company I’ve sent it to can print maps with covers, so I wrote a program to generate a cover from suitable images: the test example is [here](https://www.routemaster.app/software/images/acover.png) (10MB). And I documented my procedure [here](https://www.routemaster.app/software/mapping.html). It isn’t a path strewn with flowers, but I’ll try to take away some of the sharper edges.

The improvements I make to the original map are purely cosmetic, and perhaps in most people’s eyes I’m being a bit obsessive. They are:
- I add a title in a style suited for large format maps.
- I add a UTM grid labelled outside the map. 
- I add a lat/long graticule likewise labelled outside the map. 
- I add a scale bar below the map. 
- I add a sheet number at the bottom right.

**Comment by danielhjames** [comment]
Good job @ColinChampion ! Thanks for documenting your workflow 👍

**Comment by danielhjames** [comment]
> There's also proprietary [inkatlas](https://inkatlas.com/create/) that allows to select a specific scale

Thanks for the tip, unfortunately it doesn't output formats that can used directly in CAD software such as LibreCAD.

> Also, vanilla OpenStreetMap allows to manually set scale via the "Share" feature

True, but without a scale bar it is not easy to calibrate the maps that are produced.

**Comment by ColinChampion** [comment]
I stepped out of this discussion because it was devouring too much effort. I’d decided that being able to supply a style sheet was more critical than supplying a scale, but this may be because my own interest is in topo maps and OCitysmap is intended for town plans. What I really needed was to be able to install Mapnik on my own laptop, and then supply a bounding box, a scale, and a style sheet, and get a png as output. Formatting the map for printing would then be a relatively easy task (assuming a single-page map with no index). But I came to the conclusion that Mapnik wasn’t set up for naive users like myself.

**Comment by ColinChampion** [comment]
It might be worth mentioning one of the problems which led to my getting disheartened. You expect a topo map to identify major hills or mountains in its area. My test map included Ausangate, a 6000m mountain in southern Peru, but under the Veloroad style sheet it is neither named nor marked in any whatever – you have to work out where it is from the contours. But marking the right summits is quite a challenge. In this part of Peru it would probably be good enough to adopt a style sheet which marked all named 5000m summits, but obviously a different rule would be needed for the UK. So the user really does need to be able to supply his own style sheet; extending the set in OCitysmap would never be sufficient. That was my conclusion, at any rate.

**Comment by hholzgra** [comment]
"... extending the set in OCitysmap would never be sufficient ..."

There aren't really any styles bundled with OCitysMap at all. Well, technically there are two: "maposmatic-printable", and "pierre-orange", but the maposmatic style was never really more than a proof-of-concept, and the pierre-orange one I never even offered on the public server for reasons I don't even remember anymore.

What I offer on the public server, and in the Vagrant VM setup, is basically a collection of all somewhat usable Mapnik and CartoCSS styles published as open source that I could find.

Veloroad is a style that's mostly focused on cyclists, it is not meant to be a general purpose style, and peaks are especially of rather little interest for cyclists.

The best general topographic styles would be the OpenTopoMap, with its extra processing for saddle, viewpoint and station directions. The OpenArdenneMap style looks even better IMHO, but only when having the extra preprocessed contour/hillshading files that are only available for Belgium and not for the whole planet.

And the most useful general style overall still being the default OpenStreetMap Carto style, as pretty much everything else is either a fork of that with just some country specific adjustments, or a "special interest" style with more limited feature scope ...

Personally I'm not good at designing map styles, or at graphics design in general, I'm "just" doing the integration work and some minor adjustments here and there (e.g. converting most of the OpenTopoMap from PNG to SVG for better print results).

Long story short: the stylesheet problem will be the same with any Mapnik based render tool, it is not OCitysMap specfic.

The "can't choose map scale factor up front" on the other hand is a serious OCitysMap limitation that someone needs to tackle some day, with that someone probably being me. There's just the three usual problems with it: having the time, having the motivation, and most important: having both at the same time ...

**Comment by Talkless** [comment]
> There's just the three usual problems with it: having the time, having the motivation, and most important: having both at the same time ...

Yep, that's really understandable...

**Comment by Talkless** [comment]
> I was thinking of having the zoom being some variable somewhere while adding a slider that alters the zoom.

Mutually exclusive set of button should work, selecting  ..., 1:10000, 1:25000, 1:50000, etc, but also could be "custom" which enables slider, that would cover like 100% of use cases, whatever strange or unknown?

Personally I target about 1:20000, as it covers enough area for my hiking-with-compass trips, and yet still has enough OSM details.

Additional cool feature would be to select scale, but OVERRIDE zoom level for tiles. I.e. I would use 1:25000, but tiles as it was 1:20000. At 1:25000 I lose small details like paths...
